s3transfer/download.py

```diff
@@ -0,0 +1,311 @@
+        """Downloads an object and places content into io queue
+
+        :param client: The client to use when calling GetObject
+        :param bucket: The bucket to download from
+        :param key: The key to download from
+        :param f: The file handle to write content to
+        :param exta_args: Any extra arguements to include in GetObject request
+        :param callbacks: List of progress callbacks to invoke on download
+        :param max_attempts: The number of retries to do when downloading
+        :param io_executor: The executor that will handle the writing of
+            contents
+        :param start_index: The location in the file to start writing the
+            content of the key to.
+        """
+        last_exception = None
+        for i in range(max_attempts):
+            try:
+                response = client.get_object(
+                    Bucket=bucket, Key=key, **extra_args)
+                streaming_body = StreamReaderProgress(
+                    response['Body'], callbacks)
+
+                current_index = start_index
+                chunks = iter(
+                    lambda: streaming_body.read(self.STREAM_CHUNK_SIZE), b'')
+                for chunk in chunks:
+                    # If the transfer is done because of a cancellation
+                    # or error somewhere else, stop trying to submit more
+                    # data to be written and break out of the download.
+                    if not self._transfer_coordinator.done():
+                        io_executor.submit(
+                            IOWriteTask(
+                                self._transfer_coordinator,
+                                main_kwargs={
+                                    'f': f,
+                                    'data': chunk,
+                                    'offset': current_index
+                                }
+                            )
+                        )
+                        current_index += len(chunk)
+                    else:
+                        return
+                return
+            except S3_RETRYABLE_ERRORS as e:
+                logger.debug("Retrying exception caught (%s), "
+                             "retrying request, (attempt %s / %s)", e, i,
+                             max_attempts, exc_info=True)
+                last_exception = e
+                # Also invoke the progress callbacks to indicate that we
+                # are trying to download the stream again and all progress
+                # for this GetObject has been lost.
+                invoke_progress_callbacks(
+                    callbacks, start_index - current_index)
+                continue
+        raise RetriesExceededError(last_exception)
+
+
+class JoinFuturesTask(Task):
+    """A task to wait for the completion of other futures
+
+    :params future_to_wait_on: A list of futures to wait on
+    """
+    def _main(self, futures_to_wait_on):
+        pass
```

Comment:
It might be nice adding a comment on how this works.  I had to read through the `Task` base class to understand why this worked by just having a noop `_main`.